### PR TITLE
control-service: Add vdk-heartbeat image

### PR DESCRIPTION
**Pull request by doks5**
As part of the the CI/CD process, we could benefit from
having the vdk-heartbeat package built as a docker image.

This change adds a dockerfile to facilitate building the
vdk-heartbeat docker image when a new release of the
Control Service is made.

Testing Done: Built the image in docker locally and
ran it to verify that it is being executed as expected.

Signed-off-by: Andon Andonov <andonova@vmware.com>

### CODE REVIEW

File: projects/control-service/cicd/.gitlab-ci.yml [hunk 1]
```diff
@@ -237,3 +237,26 @@ control_service_release:
...
+    - docker:19.03.8-dind
+  variables:
+    DOCKER_HOST: tcp://localhost:2375
+    DOCKER_DRIVER: overlay2
+    DOCKER_TLS_CERTDIR: ""
+  stage: release
+  script:
+    - export CHART_NAME=pipelines-control-service
+    - export CHART_VERSION=$(cat projects/control-service/projects/helm_charts/$CHART_NAME/version.txt)
+    - export HEARTBEAT_TEST_IMAGE_NAME=$CHART_NAME-integration-tests:$CHART_VERSION
```
[the remark this answers is not in the document]
Comment (doks5) [reply projects/control-service/cicd/.gitlab-ci.yml:253]: Discussed offline.

File: projects/control-service/cicd/vdk_heartbeat_config.ini [hunk 2]
```diff
@@ -0,0 +1,13 @@
+[DEFAULT]
+# Genereate with https://console-stg.cloud.vmware.com/csp/gateway/portal/#/user/tokens
+# VDKCLI_OAUTH2_REFRESH_TOKEN= passed as environment variable
+
+JOB_RUN_TEST_MODULE_NAME=vdk.internal.heartbeat.trino_database_test
+JOB_RUN_TEST_CLASS_NAME=TrinoDatabaseRunTest
```
[the remark this answers is not in the document]
Comment (doks5) [reply projects/control-service/cicd/vdk_heartbeat_config.ini]: Fixed, thanks

File: projects/control-service/cicd/Dockerfile.vdk.heartbeat [hunk 3]
```diff
@@ -0,0 +1,22 @@
+# Install all required packages and prepare vdk heartbeat
+FROM python:3.7-slim AS builder
+
+RUN apt-get update && \
+    apt-get install -y --no-install-recommends build-essential gcc
+
+RUN python -m venv /opt/buildenv
+ENV PATH="/opt/buildenv/bin:$PATH"
+
+RUN pip install -U pip wheel quickstart-vdk && \
```
[the remark this answers is not in the document]
Comment (doks5) [reply projects/control-service/cicd/Dockerfile.vdk.heartbeat]: Discussed offline and fixed.

File: projects/control-service/cicd/vdk_heartbeat_config.ini [hunk 4]
```diff
@@ -0,0 +1,12 @@
+[DEFAULT]
+# Genereate with https://console-stg.cloud.vmware.com/csp/gateway/portal/#/user/tokens
+# VDKCLI_OAUTH2_REFRESH_TOKEN= passed as environment variable
+VDKCLI_OAUTH2_URI=https://console-stg.cloud.vmware.com/csp/gateway/am/api/auth/api-tokens/authorize
```
[the remark this answers is not in the document]
Comment (doks5) [reply projects/control-service/cicd/vdk_heartbeat_config.ini]: Discussed offline.

File: projects/control-service/cicd/start_heartbeat.sh [hunk 5]
```diff
@@ -0,0 +1,8 @@
+#!/bin/bash
+
+# Copyright 2021 VMware, Inc.
+# SPDX-License-Identifier: Apache-2.0
+
+pip install -U $VDK_HEARTBEAT_VDK_DISTRIBUTION_NAME --extra-index-url $VDK_HEARTBEAT_PIP_EXTRA_INDEX_URL
```
Comment (tozka) [inline projects/control-service/cicd/start_heartbeat.sh:6]: Set default . 

One of our goals is to make everything we do easy to get started with. That means no required configurations as much as possible in this case. We should aim for every configuration value we expose should have some sensible default -> less things for users to have to configure at first -> This way we reduce start up time for new users -> happier users. 


```
VDK_HEARTBEAT_VDK_DISTRIBUTION_NAME=${VDK_HEARTBEAT_VDK_DISTRIBUTION_NAME:-'quickstart-vdk'}
VDK_HEARTBEAT_PIP_EXTRA_INDEX_URL=${VDK_HEARTBEAT_PIP_EXTRA_INDEX_URL:-'https://pypi.org/simple'}

File: projects/control-service/cicd/.gitlab-ci.yml [hunk 6]
```diff
@@ -237,3 +237,18 @@ control_service_release:
...
+    name: gcr.io/kaniko-project/executor:debug
+    entrypoint: [""]
+  stage: release
+  script:
+    - mkdir -p /kaniko/.docker
+    - echo "{\"auths\":{\"$VDK_HEARTBEAT_REGISTRY\":{\"auth\":\"$(echo -n ${VDK_HEARTBEAT_REGISTRY_USER}:${VDK_HEARTBEAT_PASSWORD} | base64)\"}}}" > /kaniko/.docker/config.json
+    - /kaniko/executor --context ../../cicd --dockerfile Dockerfile.vdk.heartbeat --destination $CHART_NAME_integration_test:$CHART_VERSION
+  only:
+    refs:
+      - main
```
[the remark this answers is not in the document]
Comment (mivanov1988) [reply projects/control-service/cicd/.gitlab-ci.yml:260]: Did you test it?

File: projects/control-service/cicd/start_heartbeat.sh (hunk 5 above)
Comment (doks5) [reply projects/control-service/cicd/start_heartbeat.sh:9]: Done, thanks

File: projects/control-service/cicd/.gitlab-ci.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (doks5) [reply projects/control-service/cicd/.gitlab-ci.yml]: @mivanov1988 I did test it, but got "access denied" at the docker push step https://gitlab.com/vmware-analytics/versatile-data-kit/-/jobs/1648349083
I guess something's not right with the dockerhub credentials.

File: projects/control-service/cicd/.gitlab-ci.yml [hunk 7]
```diff
@@ -237,3 +237,27 @@ control_service_release:
...
+    - docker:19.03.8-dind
+  variables:
+    DOCKER_HOST: tcp://localhost:2375
+    DOCKER_DRIVER: overlay2
+    DOCKER_TLS_CERTDIR: ""
+  stage: build
+  script:
+    - export CHART_NAME=pipelines-control-service
+    - export CHART_VERSION=$(cat projects/control-service/projects/helm_charts/$CHART_NAME/version.txt)
+    - export VDK_HEARTBEAT_TEST_IMAGE_NAME=$CHART_NAME-integration-tests:0.0.1
```
[the remark this answers is not in the document]
Comment (mivanov1988) [reply projects/control-service/cicd/.gitlab-ci.yml:253]: Probably it is left over from manual tests.

[the remark this answers is not in the document]
Comment (doks5) [reply projects/control-service/cicd/.gitlab-ci.yml:253]: This is only for testing purposes. Will return CHART_VERSION when finished testing.

File: projects/control-service/cicd/.gitlab-ci.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (mivanov1988) [reply projects/control-service/cicd/.gitlab-ci.yml]: I guess the image prefix is not correct - 
`export VDK_HEARTBEAT_IMAGE_NAME="$VDK_DOCKER_REGISTRY_URL/$VDK_HEARTBEAT_TEST_IMAGE_NAME"`

[the remark this answers is not in the document]
Comment (mivanov1988) [reply projects/control-service/cicd/.gitlab-ci.yml]: Check the login command locally.

[the remark this answers is not in the document]
Comment (doks5) [reply projects/control-service/cicd/.gitlab-ci.yml]: Fixed it. It's working now. https://gitlab.com/vmware-analytics/versatile-data-kit/-/jobs/1649284408

File: projects/control-service/cicd/.gitlab-ci.yml [hunk 8]
```diff
@@ -237,3 +237,28 @@ control_service_release:
...
+    DOCKER_TLS_CERTDIR: ""
+  stage: release
+  script:
+    - export CHART_NAME=pipelines-control-service
+    - export CHART_VERSION=$(cat projects/control-service/projects/helm_charts/$CHART_NAME/version.txt)
+    - export VDK_HEARTBEAT_TEST_NAME=$CHART_NAME-integration-tests:$CHART_VERSION
+    - export VDK_HEARTBEAT_IMAGE_NAME="$VDK_DOCKER_REGISTRY_URL/$VDK_HEARTBEAT_TEST_NAME"
+    - docker login $VDK_DOCKER_REGISTRY_URL --username $VDK_DOCKER_REGISTRY_USERNAME --password $VDK_DOCKER_REGISTRY_PASSWORD
+    - cd projects/control-service/cicd
+    - docker build --tag $VDK_HEARTBEAT_IMAGE_NAME -f Dockerfile.vdk.heartbeat .
```
Comment (tozka) [inline projects/control-service/cicd/.gitlab-ci.yml:257]: Why not use Kaniko? I am not suggesting changing it. Using docker is fine. But I would like to understand better what limitations or issues Kaniko had that we switched to docker.

Comment (doks5) [reply projects/control-service/cicd/.gitlab-ci.yml:257]: We have an example with docker-in-docker in the ci/cd script, and we weren't sure if kaniko would work pushing the image to docker hub. That's why dind was used, just to be on the safe side.

**Review by tozka — APPROVED** [review]

**Pull request merged**